### PR TITLE
Add new devices, fractional trains etc.

### PR DESCRIPTION
**Pull request by MilhouseVH**
This PR adds

1) missing devices (project maintainers: please check errors/ommissions/improvements)
 - [x] Amlogic @chewitt 
 - [x] Allwinner @jernejsk 
 - [x] NXP/Qualcomm @lrusak 
 - [ ] Rockchip @Kwiboo 
 - [x] RPi/Generic @MilhouseVH 
2) support for "fractional trains" ie. 9.2, 9.4 and not just 9.0, 10.0
3) implements the concept of "unstable" builds for odd-numbered releases (unstable releases will be announced in the next even/stable train, ie. 9.1.003 will be announced in 9.2)
4) adds arguments:
```
usage: releases.py [-h] -i DIRECTORY -u URL [-o DIRECTORY]

Update LibreELEC releases.json with available tar/img.gz files.

optional arguments:
  -h, --help             show this help message and exit
  -i DIRECTORY, --input DIRECTORY
                         Directory to parsed (release files, and any existing
                         releases.json). By default releases.json will be written into
                         this directory. Required property.
  -u URL, --url URL      Base URL for releases.json. Required property.
  -o DIRECTORY, --output DIRECTORY
                         Optional directory into which releases.json will be written.
                         Defaults to same directory as --input.
```
5) output is sorted for consistent results
6) output is formatted for readability/editing
7) uboot images now appear in a `uboot[]` array

Tested with Python2 and Python3.

### CODE REVIEW

**Comment by chewitt** [comment]
Output generated by this PR is currently live on the releases server so you can check content via USB/SD Creator and/or the settings add-on in manual update mode. LGTM.

**Comment by CvH** [comment]
RK and likely Allwinner are currently semi broken at the USB Tool,you need to check "show all" to have anything to choose from. Likely the USB tool misses some support for that style.

File: releases.py [hunk 1]
```diff
@@ -26,36 +63,63 @@ def file_as_blockiter(afile, blocksize=65536):
...
+                             'Dragonboard.arm': 'Qualcomm Dragonboard',
+                             'Generic.x86_64': 'Generic AMD/Intel/NVIDIA (x86_64)',
+                             'H3.arm': 'Allwinner H3',
+                             'H6.arm': 'Allwinner H6',
                              'imx6.arm': 'Cubox-i2/i4 and Hummingboard (iMX6)',
-                             'KVIM.arm': 'Khadas Vim',
-                             'KVIM2.arm': 'Khadas Vim 2',
-                             'LePotato.arm': 'LePotato (AML-S905X-CC)',
-                             'MiQi.arm': 'Rockchip MiQi RK3328',
+                             'iMX6.arm': 'Cubox-i2/i4 and Hummingboard (iMX6)',
```
Comment (InuSasha) [inline releases.py]: two iMX6 in different cases intentional?

Comment (MilhouseVH) [reply releases.py]: Yes, old imx6 and new iMX6... :man_shrugging:

**Comment by chewitt** [comment]
> RK and likely Allwinner are currently semi broken at the USB Tool,you need to check "show all" to have anything to choose from. Likely the USB tool misses some support for that style.

Nothing broken, USB/SD only shows stable releases unless 'show all' is checked and neither RK or Allwinner have stable releases yet.

**Comment by sky42src** [comment]
I do like the new versioning scheme, but i did not fully understand it.
what would be the version number for the 9.2 Beta 1, so that one could see (or regex) this is 9.2 Beta 1?

**Comment by sky42src** [comment]
Feature Request: support for BUILDER_NAME in image file name and create extra trains per builder. my https://github.com/LibreELEC/LibreELEC.tv/pull/3798 is the part for the build system.

Feature Request: a train suffix to be optional given on the command line, like --suffix test would create e.g. LibreELEC-9.2-test

**Comment by chewitt** [comment]
If you remove imx6.arm from the list the script will error instead of not including the imx6.arm files which exist on the server - so it needs to be added back. If you keep matching descriptions both old/new filenames show up under the same description. NB: Script failure due to undocumented new files is desirable else people won't notice that something new was added.

**Comment by sky42src** [comment]
The last time i played around with releases.json was with 8.2.5 and the upcoming 9.0 and i found
another reason why a builder releases.json should have different trains then then the official one.
When you have the official releases.json and custom url 1 and 2.
They all 3 have a train named "LibreELEC-9.0" for official, custom1 and custom2, then the last one wins (custom2) and you get from custom2. I am not able to install from LibreELEC-9.0.
I did not recheck that lateley.

**Comment by MilhouseVH** [comment]
>  what would be the version number for the 9.2 Beta 1, so that one could see (or regex) this is 9.2 Beta 1?

9.2beta1 isn't really possible (given the current scheme) - we can only use the unstable 9.1.x series for the various 9.2 "alpha/beta" releases. At some point the value of x may denote "beta1", this could either be the next sequential value after the previous alpha (ie. 9.1.004) or we could sub-divide x into alpha/beta groups, ie. alpha 000-500 and beta 501-999 (or similar - 3 digits is excessive).

However I think we should really try to avoid mid-release alpha/beta's, it's confusing and also indicative of bad practice. When introducing a new device mid-release (as we did with RPi4) we should isolate code changes so that they are device specific ensuring that all other devices are unaffected. There should not be a need for alpha/beta releases, and when necessary we can release a series of 9.1.x, 9.3.x etc. releases as "preview" releases limited (ideally) to the new device.

Mentioning alpha/beta (as we did with the 9.1.x releases) gives the impression the quality of the core code base may have regressed in some significant way, which should not normally be the case - if that does happen then we should be considering a new major release (ie. 10.0).

In short I think we need to rethink how we release new devices mid-release in future, as we should reserve alpha/beta for use during a new major release cycle.

**Comment by sky42src** [comment]
I need to extend/change the wish for Buildername in trains. If that will be supported, there should also be a command line option to set the builder name for all file where none is detected in the filename. This is for old image files, to not all change them. For me it would be also enough to have the command line option and not detecting it in file names.

**Comment by MilhouseVH** [comment]
> If you remove imx6.arm from the list the script will error instead of not including the imx6.arm files which exist on the server

I have re-introduced the legacy imx6 entry.

**Comment by MilhouseVH** [comment]
@sky42src I've added builder support based on https://github.com/LibreELEC/LibreELEC.tv/pull/3798 - please confirm if this now produces what you require.

@chewitt I've also added a change which will remove any builds without releases.

**Comment by sky42src** [comment]
@MilhouseVH Thank you very much for that builder support feature. I changed my setup and used the new releases.py for all 9.x and i let my old 8.2 as they are.
I do have a different pretty name regex like this ([0-9]+\.[0-9]+\.[0-9]+\.[0-9]+-?[0-9a-z#]*). To overwrite the prettyname regex i would love a command line option.

My new releases.json are here
http://sky42.libreelec.tv/
each dir over here http://sky42.libreelec.tv/release/ (8.2 is old with my old changed script)
and for testing http://sky42.libreelec.tv/testing/.next/
just created and tested them and it is excellent.

Thank you!

**Comment by MilhouseVH** [comment]
@sky42src I've added support for a custom prettyname regex via the `-p/--prettyname` option.

**Comment by sky42src** [comment]
@MilhouseVH works perfect
Thank you!

**Comment by sky42src** [comment]
How about not calculating the sha256 in the releases.py when it already exist, as a sha256 file, and just read it and write that one in releases.json?

**Comment by chewitt** [comment]
Because the sha256 files do not exist on the server, thus there is nothing to read hashes from.

**Comment by sky42src** [comment]
Oh i see.
On my server they are and i hide them with `fancyindex_ignore \.sha256$;`

**Comment by sky42src** [comment]
Tested the the UBOOT commit and the releases.json files are here
http://sky42.libreelec.tv/releases.json
http://sky42.libreelec.tv/release/9.0/releases.json
http://sky42.libreelec.tv/release/9.2/releases.json

but i dont have any RK devices and can not test if and how it works.

**Comment by MilhouseVH** [comment]
@sky42 the uboot entries in http://sky42.libreelec.tv/releases.json don't look to be correct - I'll look into it...

**Comment by MilhouseVH** [comment]
@sky42 OK not to worry, it's just Firefox truncating the uboot image name when the node isn't expanded:

s1

The name is shown correctly when the node is expanded.

Note that you have these files [here](http://sky42.libreelec.tv/.channels/release/all/):

```
LibreELEC-Generic.x86_64-sky42-9.1.501-#190907.tar
LibreELEC-Generic.x86_64-sky42-9.1.501-#190907.img.gz
LibreELEC-Generic.x86_64-sky42-9.1.501-#190907-kernel-4.19.tar
LibreELEC-Generic.x86_64-sky42-9.1.501-#190907-kernel-4.19.img.gz
```

which results in the img.gz file with `-kernel` suffix being interpreted as a uboot image for the tar file without the suffix:

```
          "1": {
            "file": {
              "name": "LibreELEC-Generic.x86_64-sky42-9.1.501-#190907-kernel-4.19.tar",
              "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
              "size": "0"
            },
            "image": {
              "name": "LibreELEC-Generic.x86_64-sky42-9.1.501-#190907-kernel-4.19.img.gz",
              "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
              "size": "0"
            }
          },
          "2": {
            "file": {
              "name": "LibreELEC-Generic.x86_64-sky42-9.1.501-#190907.tar",
              "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
              "size": "0"
            },
            "image": {
              "name": "LibreELEC-Generic.x86_64-sky42-9.1.501-#190907.img.gz",
              "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
              "size": "0"
            },
            "uboot": [
              {
                "name": "LibreELEC-Generic.x86_64-sky42-9.1.501-#190907-kernel-4.19.img.gz",
                "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
                "size": "0"
              }
            ]
          }
```

**Comment by MilhouseVH** [comment]
@sky42 The `uboot[]` array is only used by the USB/SD Creator which I doubt your users are using, so bogus testing entries being misinterpreted as a uboot image by this script shouldn't be a problem for LibreELEC clients as the settings add-on will ignore the u-boot information entirely.

**Comment by sky42src** [comment]
@MilhouseVH works for me and a little regex before and after releases.py renames the trains to avoid same names as official releases.

here are the files
http://sky42.libreelec.tv/releases.json
http://sky42.libreelec.tv/release/9.0/releases.json
http://sky42.libreelec.tv/release/9.2/releases.json
http://sky42.libreelec.tv/testing/.next/releases.json

and here is how i create them (not pretty, but works)
http://sky42.libreelec.tv/.channels/create-releases-json.sh
http://sky42.libreelec.tv/testing/.next/create-releases-json.sh

**Comment by lrusak** [comment]
Did anything ever come about with this? It never merged but is it currently used?

We need to update the script to allow using the device suffixed images for use with the usb-sd-creator (which needs changes also).
